colexec: a flake in TestExternalSort

On a5d61d2cf2bfc11bd6bdc27dcab32467334fb123:
```
./dev test pkg/sql/colexec -v --stream-output --ignore-cache -f=TestExternalSort\$\$ -- --crdb_test_off --test_env=COCKROACH_RANDOM_SEED=9091000296315083269
```
```
I230411 22:47:40.851977 19 sql/colexec/external_sort_test.go:75  [-] 461  spillForced=true/numRepartitions=3/k < input length
I230411 22:47:40.851983 19 sql/colexec/colexectestutils/utils.go:691  [-] 462  batchSize=1/sel=false
I230411 22:47:40.852362 19 sql/colexec/colexectestutils/utils.go:537  [-] 463  reusing after reset
    utils.go:358: expected didn't match actual. diff:
        @@ -1,4 +1,4 @@
        -0: [1]
        -1: [2]
        -2: [3]
        +0: [2]
        +1: [3]
        +2: [4]
    panic.go:522: -- test log scope end --
test logs left over in: /Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/1dc11c62c6f7bfe6fd9ffb420f09fb85/logTestExternalSort1466902768
--- FAIL: TestExternalSort (0.11s)
```

Jira issue: CRDB-26885

--- a/pkg/sql/colexec/sorttopk.eg.go
+++ b/pkg/sql/colexec/sorttopk.eg.go
@@ -173,6 +173,9 @@ func spool_true(t *topKSorter) {
 				groupId = __retval_groupId
 			}
 		}
+		// Importantly, the memory limit can only be reached _after_ the tuples
+		// are appended to topK, so all fromLength tuples are considered
+		// "processed".
 		t.firstUnprocessedTupleIdx = fromLength
 		t.topK.AppendTuples(t.inputBatch, 0 /* startIdx */, fromLength)
 		remainingRows -= uint64(fromLength)
@@ -211,8 +214,8 @@ func spool_true(t *topKSorter) {
 					{
 						var __retval_groupDone bool
 						{
-							for i := t.firstUnprocessedTupleIdx; i < t.inputBatch.Length(); i++ {
-								idx := sel[i]
+							for ; t.firstUnprocessedTupleIdx < t.inputBatch.Length(); t.firstUnprocessedTupleIdx++ {
+								idx := sel[t.firstUnprocessedTupleIdx]
 								// If this is a distinct group, we have already found the top K input,
 								// so we can stop comparing the rest of this and subsequent batches.
 								if t.orderState.distinctOutput[idx] {
@@ -232,7 +235,6 @@ func spool_true(t *topKSorter) {
 									heap.Fix(t.heaper, 0)
 								}
 							}
-							t.firstUnprocessedTupleIdx = t.inputBatch.Length()
 							{
 								__retval_groupDone = false
 							}
@@ -244,8 +246,8 @@ func spool_true(t *topKSorter) {
 					{
 						var __retval_groupDone bool
 						{
-							for i := t.firstUnprocessedTupleIdx; i < t.inputBatch.Length(); i++ {
-								idx := i
+							for ; t.firstUnprocessedTupleIdx < t.inputBatch.Length(); t.firstUnprocessedTupleIdx++ {
+								idx := t.firstUnprocessedTupleIdx
 								// If this is a distinct group, we have already found the top K input,
 								// so we can stop comparing the rest of this and subsequent batches.
 								if t.orderState.distinctOutput[idx] {
@@ -265,7 +267,6 @@ func spool_true(t *topKSorter) {
 									heap.Fix(t.heaper, 0)
 								}
 							}
-							t.firstUnprocessedTupleIdx = t.inputBatch.Length()
 							{
 								__retval_groupDone = false
 							}
@@ -328,6 +329,9 @@ func spool_false(t *topKSorter) {
 			// t.topK will be full after this batch.
 			fromLength = int(remainingRows)
 		}
+		// Importantly, the memory limit can only be reached _after_ the tuples
+		// are appended to topK, so all fromLength tuples are considered
+		// "processed".
 		t.firstUnprocessedTupleIdx = fromLength
 		t.topK.AppendTuples(t.inputBatch, 0 /* startIdx */, fromLength)
 		remainingRows -= uint64(fromLength)
@@ -361,8 +365,8 @@ func spool_false(t *topKSorter) {
 			func() {
 				if sel != nil {
 					{
-						for i := t.firstUnprocessedTupleIdx; i < t.inputBatch.Length(); i++ {
-							idx := sel[i]
+						for ; t.firstUnprocessedTupleIdx < t.inputBatch.Length(); t.firstUnprocessedTupleIdx++ {
+							idx := sel[t.firstUnprocessedTupleIdx]
 							maxIdx := t.heap[0]
 							groupMaxIdx := 0
 							if compareRow_false(t, inputVecIdx, topKVecIdx, idx, maxIdx, groupId, groupMaxIdx) < 0 {
@@ -372,12 +376,11 @@ func spool_false(t *topKSorter) {
 								heap.Fix(t.heaper, 0)
 							}
 						}
-						t.firstUnprocessedTupleIdx = t.inputBatch.Length()
 					}
 				} else {
 					{
-						for i := t.firstUnprocessedTupleIdx; i < t.inputBatch.Length(); i++ {
-							idx := i
+						for ; t.firstUnprocessedTupleIdx < t.inputBatch.Length(); t.firstUnprocessedTupleIdx++ {
+							idx := t.firstUnprocessedTupleIdx
 							maxIdx := t.heap[0]
 							groupMaxIdx := 0
 							if compareRow_false(t, inputVecIdx, topKVecIdx, idx, maxIdx, groupId, groupMaxIdx) < 0 {
@@ -387,7 +390,6 @@ func spool_false(t *topKSorter) {
 								heap.Fix(t.heaper, 0)
 							}
 						}
-						t.firstUnprocessedTupleIdx = t.inputBatch.Length()
 					}
 				}
 			},

--- a/pkg/sql/colexec/sorttopk.go
+++ b/pkg/sql/colexec/sorttopk.go
@@ -189,6 +189,8 @@ func (t *topKSorter) Reset(ctx context.Context) {
 	t.firstUnprocessedTupleIdx = 0
 	t.topK.ResetInternalBatch()
 	t.emitted = 0
+	t.exportedFromTopK = 0
+	t.exportedFromBatch = 0
 	if t.hasPartialOrder {
 		t.orderState.distincter.(colexecop.Resetter).Reset(t.Ctx)
 	}

--- a/pkg/sql/colexec/sorttopk_tmpl.go
+++ b/pkg/sql/colexec/sorttopk_tmpl.go
@@ -70,11 +70,11 @@ func processGroupsInBatch(
 func processBatch(
 	t *topKSorter, groupId int, sel []int, partialOrder bool, useSel bool,
 ) (groupDone bool) {
-	for i := t.firstUnprocessedTupleIdx; i < t.inputBatch.Length(); i++ {
+	for ; t.firstUnprocessedTupleIdx < t.inputBatch.Length(); t.firstUnprocessedTupleIdx++ {
 		if useSel {
-			idx := sel[i]
+			idx := sel[t.firstUnprocessedTupleIdx]
 		} else {
-			idx := i
+			idx := t.firstUnprocessedTupleIdx
 		}
 		if partialOrder {
 			// If this is a distinct group, we have already found the top K input,
@@ -98,7 +98,6 @@ func processBatch(
 			heap.Fix(t.heaper, 0)
 		}
 	}
-	t.firstUnprocessedTupleIdx = t.inputBatch.Length()
 	return false
 }
 
@@ -141,6 +140,9 @@ func spool(t *topKSorter, partialOrder bool) {
 				groupId = processGroupsInBatch(t, fromLength, groupId, sel, false)
 			}
 		}
+		// Importantly, the memory limit can only be reached _after_ the tuples
+		// are appended to topK, so all fromLength tuples are considered
+		// "processed".
 		t.firstUnprocessedTupleIdx = fromLength
 		t.topK.AppendTuples(t.inputBatch, 0 /* startIdx */, fromLength)
 		remainingRows -= uint64(fromLength)

